src/artifacts.js

```diff
@@ -423,6 +643,117 @@ var replyWithArtifact = async function(taskId, runId, name, req, res) {
   this.monitor.reportError(err);
 };
 
+/** Complete artifact */
+api.declare({
+  method:     'put',
+  route:      '/task/:taskId/runs/:runId/artifacts/:name(*)',
+  name:       'completeArtifact',
+  stability:  API.stability.experimental,
+  scopes: [
+    [
+      'queue:create-artifact:<name>',
+      'assume:worker-id:<workerGroup>/<workerId>',
+    ], [
+      'queue:create-artifact:<taskId>/<runId>',
+    ],
+  ],
+  deferAuth:  true,
+  input:      'put-artifact-request.json#',
+  title:      'Complete Artifact',
+  description: 'tbd',
+}, async function(req, res) {
+  let taskId      = req.params.taskId;
+  let runId       = parseInt(req.params.runId, 10);
+  let name        = req.params.name;
+  let input       = req.body;
+  let isPublic    = /^public\//.test(name);
+
+  let [artifact, task] = await Promise.all([
+    this.Artifact.load({taskId, runId, name}, true),
+    this.Task.load({taskId}, true),
+  ]);
+
+  let run = task.runs[runId];
+  if (!run) {
+    return res.reportError('InputError',
+      'Run not found',
+      {});
+  }
+  let workerGroup = run.workerGroup;
+  let workerId = run.workerId;
+
```

Comment:
@jhford, we should return 409, `RequestConflict` if the run have been resolved...

Specifically, if `run.state` is not `running` or `exception` (with `run.resolved` within the past 25 minutes).
Same condition we have for creating artifacts:
https://github.com/taskcluster/taskcluster-queue/blob/9b36857aa8f8f27515382b48e638e9754ebf5993/src/artifacts.js#L160-L181